api/v1/logstorage_types.go

```diff
@@ -52,6 +52,10 @@ type LogStorageSpec struct {
 	// Only ECKOperator is supported for this spec.
 	// +optional
 	ComponentResources []LogStorageComponentResource `json:"componentResources,omitempty"`
+
+	// IngestReplicaCount can be used to adjust the replica count for ES Gateway.
+	// +optional
+	IngestReplicaCount *int32 `json:"ingestReplicaCount,omitempty"`
```

Comment:
I feel a bit conflicted about this field. If it is used, are we then supposed to not have Nodes, Indices, etc, since ES is outsourced to an external cluster? I would expect to see validation to ensure this is the case.

I also feel that the name IngestReplicaCount can easily lead to confusions, since ES also has a concept of ingestion nodes. I would be in favor of adding an es-gateway struct, which then has more clarification around it's setup (replica's in this case), to clearly divide regular logstorage from the es-gateway scenario.

Reply:
@rene-dekker, @tmjd had request Vlad change this. @rene-dekker is right that ingest is an overloaded term here, and I would like to reserve it for the specific elasticsearch node type, should we decide to optimise the Elasticsearch clusters and allow users to configure it.

@tmjd what was the reasoning you don't want to expose that there's a gateway in these settings? I'd agree that if there was a way to deduce the number of replicas for the gateway that we shouldn't expose it, but that's not possible, so I would argue that it should be clear that the setting changes the gateway replica count (Gateway section with ReplicaCount). The resource will be called a gateway in the cluster, no reason to hide a user facing setting for it.

Reply:
I don't want to expose this to the user to because I believe it is something that user shouldn't specifically be aware of. For example, if we moved this functionality to voltron then a Gateway setting would be a misnomer then.

Have we considered deducing the number of replicas needed for the gateway? You say it isn't possible but I think there are several ways we could address that. What if we exposed an endpoint on the gateway that the operator queried and scaled the gateway up when a certain threshold was reached.
Another option would be to use node counts to determine its scale, though we'd need to add some functionality to get node counts for managed clusters.
Another option is to scale with ES nodes, we know ES will have to be able to handle everything going through the gateway so if we know 1 gateway can saturate one ES node then that's an easy scaling.

Also we don't know how much this component will need to scale, I think we should deploy at least 2 for redundancy/HA so those 2 may be enough to scale.

Originally I believe there was no plan to expose an option for gateway scaling until @Thakurvaibhav requested it, so we had it for internal use, if the expectation is still that this would be for internal use then since we don't have a good design for this configuration I'd prefer not to expose it generally.

Reply:
I've extracted this field to a separate branch for now and we can continue to assess how we want to go about scaling the gateway after performance testing it.

Reply:
> Also we don't know how much this component will need to scale, I think we should deploy at least 2 for redundancy/HA so those 2 may be enough to scale.

Ya let's do this, that's a good plan.

Regarding everything else you said, I think you're right. I'm forgetting that this really is just a simple proxy, if it couldn't handle 100 times the load of Elasticsearch then the implementation of the gateway is extremely flawed, so 2 replicas for 3 nodes (which is the max anybody has) is _**very**_ generous.

> I've extracted this field to a separate branch for now and we can continue to assess how we want to go about scaling the gateway after performance testing it.

@vberezny if performance testing shows that the gateway is the bottleneck here then I would be very hesitant to say the way forward would be to add more gateways, I'd say the implementation should be scrutinised and figure out why it's the bottleneck and not Elasticsearch.